### PR TITLE
Image request

### PR DESCRIPTION
**Pull request by grace000**
This PR covers this [story](https://trello.com/c/8qQr7YcX/28-serving-image-file-475).

Client can go to path /kitty_image for a cute cat pic.

A few issues that I'm concerned about:

- Image requests outside of jpegs: Right now I'm only considering jpeg images, but in the future I will need to handle other types such as png and gif.
- Formatting the response before sending to the outputstream: I removed the printwriter in order to write bytes directly to the stream. However I plan on refactoring the formatting implementation so that the responsibility is no longer on the response writer.
- Headers: I'm currently collecting headers, but I'm not using them beyond collecting content length in order to parse the body. However, I plan to attach them to the response.

### CODE REVIEW

File: src/main/java/github/grace000/erversay/Request/Request.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package github.grace000.erversay.Request;
+
+import java.util.HashMap;
+
+public class Request {
+    public String method;
+    public String path;
+    public String body;
```
Comment (bgerstle) [inline src/main/java/github/grace000/erversay/Request/Request.java:8]: Probably not a blocker, but now that your `Response` body is in bytes the `Request` should eventually do the same.

File: src/main/java/github/grace000/erversay/RouteHandlers/KittyImage.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            }
+        return fileResponse.build();
+        }
+        else return responseBuilder
+                .withHeaders(OPTIONS_HEADER)
+                .withStatus(NOT_ALLOWED_STATUS)
+                .build();
+    }
+
+    public byte[] getFilePath() throws IOException {
```
Comment (bgerstle) [inline src/main/java/github/grace000/erversay/RouteHandlers/KittyImage.java]: This method name is confusing. `"public/tuxedo.jpg"` is the path, but you're returning the data of the file at that path. You should consider naming this something like `readFileData()` which conveys not just that you're returning file data, but also that it's being read (i.e. IO is happening).

File: src/main/java/github/grace000/erversay/Constants/Headers.java [hunk 3]
```diff
@@ -6,4 +6,5 @@
     public final static String CONTENT_LENGTH = "Content-Length";
     public final static String NOT_ALLOWED_HEADER = "Allow: HEAD, OPTIONS";
     public final static String REDIRECT_HEADER = "Location: http://127.0.0.1:5000/simple_get";
+    public final static String JPEG_IMAGE_HEADER = "Content-Type: image/jpeg";
```
Comment (ssunday) [inline src/main/java/github/grace000/erversay/Constants/Headers.java:9]: So this works just for `jpeg`—what about `jpg`?

File: src/main/java/github/grace000/erversay/Request/Request.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package github.grace000.erversay.Request;
```
Comment (ssunday) [inline src/main/java/github/grace000/erversay/Request/Request.java:1]: why the nesting if it's the same word?

File: src/main/java/github/grace000/erversay/Response/ResponseWriter.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        try {
+            byte[] status = (DEFAULT_VERSION + SP + response.status + CRLF).getBytes();
+            byte[] headers = (response.headers).getBytes();
+
+            outputStream.write(status);
+            outputStream.write(headers);
+            outputStream.write((DOUBLE_LINE_FEED).getBytes());
+            outputStream.write(response.body);
+
+            System.out.println("WRITER: MESSAGE SENT");
```
Comment (ssunday) [inline src/main/java/github/grace000/erversay/Response/ResponseWriter.java]: Do you want the logging still in here?

File: src/main/java/github/grace000/erversay/RouteHandlers/KittyImage.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    public Response handle(Request request) {
+        ResponseBuilder fileResponse = responseBuilder;
+        if (isMethodAllowed(request.method)) {
+        try {
+            byte[] data = getFilePath();
+            fileResponse
+                    .withStatus(OK_STATUS)
+                    .withBody(data)
+                    .withHeaders(JPEG_IMAGE_HEADER);
```
Comment (bgerstle) [inline src/main/java/github/grace000/erversay/RouteHandlers/KittyImage.java:42]: Any request or response with a non-empty body must have a `Content-Length` header along with a `Content-Type`. I recommend using a `Map` for response headers the same you do for request headers.

File: src/main/java/github/grace000/erversay/RouteHandlers/NotAllowed.java [hunk 7]
```diff
@@ -24,11 +23,12 @@ public boolean isMethodAllowed(String method) {
...
     }
 
-    public String handle(Request request) {
+    public Response handle(Request request) {
         if (isMethodAllowed(request.method)) {
             return responseBuilder
                     .build();
-        } else return responseBuilder
+        }
+        else return responseBuilder
```
Comment (bgerstle) [inline src/main/java/github/grace000/erversay/RouteHandlers/NotAllowed.java]: Please also wrap `if` and `else` clauses in curly brackets. [Bad things](https://blog.codecentric.de/en/2014/02/curly-braces/) happen to those who don't. Also, you should consider using [guard statements](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html) to return early when given bad input.

**Review by bgerstle — CHANGES_REQUESTED** [review]
Some preliminary feedback. Any reason you aren't adding a `Content-Length` for your responses w/ bodies?

File: src/main/java/github/grace000/erversay/Constants/Headers.java (hunk 3 above)
Comment (grace000) [reply src/main/java/github/grace000/erversay/Constants/Headers.java:9]: I read that jpeg covers jpg and jpeg. I'm currently serving a jpg with this mime type.

File: src/main/java/github/grace000/erversay/Request/Request.java (hunk 4 above)
Comment (grace000) [reply src/main/java/github/grace000/erversay/Request/Request.java:1]: I'd been doing some file restructuring. I'll rename that folder.

File: src/main/java/github/grace000/erversay/Response/ResponseWriter.java (hunk 5 above)
Comment (grace000) [reply src/main/java/github/grace000/erversay/Response/ResponseWriter.java]: I don't need it. It helps me see the extra pings coming through, but I'll delete.

**Comment by grace000** [comment]
> Some preliminary feedback. Any reason you aren't adding a `Content-Length` for your responses w/ bodies?

I plan on adding the a content length to the map of headers and then sending that as part of the response object.

**Pull request merged**